monai/data/wsi_datasets.py

```diff
@@ -0,0 +1,116 @@
+# Copyright (c) MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Callable, Dict, List, Optional, Tuple, Union
+
+import numpy as np
+
+from monai.data import Dataset
+from monai.data.wsi_reader import WSIReader
+from monai.utils import ensure_tuple_rep
+
+__all__ = ["PatchWSIDataset"]
+
+
+class PatchWSIDataset(Dataset):
+    """
+    This dataset extracts patches from whole slide images (without loading the whole image)
+    It also reads labels for each patch and provides each patch with its associated class labels.
+
+    Args:
+        data: the list of input samples including image, location, and label (see the note below for more details).
+        size: the size of patch to be extracted from the whole slide image.
+        level: the level at which the patches to be extracted (default to 0).
+        transform: transforms to be executed on input data.
+        reader_name: the name of library to be used for loading whole slide imaging, as the backend of `monai.data.WSIReader`
+            Defaults to CuCIM.
+
+    Note:
+        The input data has the following form as an example:
+
+        .. code-block:: python
+
+            [
+                {"image": "path/to/image1.tiff", "location": [200, 500], "label": 0},
+                {"image": "path/to/image2.tiff", "location": [100, 700], "label": 1}
+            ]
+
+    """
+
+    def __init__(
+        self,
+        data: List,
+        size: Optional[Union[int, Tuple[int, int]]] = None,
+        level: Optional[int] = None,
+        transform: Optional[Callable] = None,
+        reader_name: str = "cuCIM",
```

Comment:
Hi @Nic-Ma, we can do that but I don't see the benefit of this generalization since this works only with a very specific readers, namely `OpenSlide` and `CuCIM`.

Reply:
Hi @drbeh ,

I think the main benefit is to support customized readers, for example, users may extend your `CuCIMReader` with some their own logic, then use it in the `PatchWSIDataset` dataset.
What do you think?

Thanks.

Reply:
Sure, I can add  (1) backend name, (3) reader class, and (4) reader instance if you think that someone would use it.
However, due to the way `WSIReader` is designed as an interface for all the concrete implementation of `BaseWSIReader` (2) reader class module path will never be used.

Reply:
Sure, sounds good to me.
If we have the request to config the expected class as `string` type in JSON config, we may need to support that later.

Thanks.

Reply:
Sure, make sense.